File: go-server/internal/auth/atproto_oauth.go
package auth

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"time"
)

// AtProtoOAuthClient handles AT Protocol OAuth flows with DPoP
type AtProtoOAuthClient struct {
	clientID   string
	httpClient *http.Client
}

// NewAtProtoOAuthClient creates a new AT Protocol OAuth client
func NewAtProtoOAuthClient(clientID string) *AtProtoOAuthClient {
	return &AtProtoOAuthClient{
		clientID: clientID,
		httpClient: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

// OAuthServerMetadata represents OAuth server metadata from .well-known discovery
type OAuthServerMetadata struct {
	Issuer                                string   `json:"issuer"`
	AuthorizationEndpoint                 string   `json:"authorization_endpoint"`
	TokenEndpoint                         string   `json:"token_endpoint"`
	PushedAuthorizationRequestEndpoint    *string  `json:"pushed_authorization_request_endpoint,omitempty"`
	DPoPSigningAlgValuesSupported         []string `json:"dpop_signing_alg_values_supported,omitempty"`
	ResponseTypesSupported                []string `json:"response_types_supported"`
	GrantTypesSupported                   []string `json:"grant_types_supported"`
	CodeChallengeMethodsSupported         []string `json:"code_challenge_methods_supported"`
}

// PARResponse represents a Pushed Authorization Request response
type PARResponse struct {
	RequestURI string `json:"request_uri"`
	ExpiresIn  int64  `json:"expires_in"`
}

// DiscoverMetadata discovers OAuth server metadata for a PDS
func (c *AtProtoOAuthClient) DiscoverMetadata(ctx context.Context, pdsURL string) (*OAuthServerMetadata, error) {
	metadataURL := fmt.Sprintf("%s/.well-known/oauth-authorization-server", pdsURL)

	// Create request with timeout context
	timeoutCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
	defer cancel()

	req, err := http.NewRequestWithContext(timeoutCtx, "GET", metadataURL, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		// Provide more detailed error message
		if timeoutCtx.Err() == context.DeadlineExceeded {
			return nil, fmt.Errorf("OAuth metadata discovery timed out after 10 seconds. The server at %s may not support AT Protocol OAuth yet, or the endpoint is unreachable. Please use --method password instead", pdsURL)
		}
		return nil, fmt.Errorf("failed to fetch OAuth metadata from %s: %w. The server may not support AT Protocol OAuth yet. Please try --method password instead", metadataURL, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("OAuth metadata discovery failed with status %d from %s. Response: %s. The server may not support AT Protocol OAuth yet. Please try --method password instead", resp.StatusCode, metadataURL, string(body))
	}

	var metadata OAuthServerMetadata
	if err := json.NewDecoder(resp.Body).Decode(&metadata); err != nil {
		return nil, fmt.Errorf("failed to decode OAuth metadata from %s: %w", metadataURL, err)
	}

	return &metadata, nil
}

// SendPAR sends a Pushed Authorization Request with DPoP
func (c *AtProtoOAuthClient) SendPAR(
	ctx context.Context,
	metadata *OAuthServerMetadata,
	pkce *PKCEParams,
	dpopKeyPair *DPoPKeyPair,
	handle string,
) (*PARResponse, error) {
	if metadata.PushedAuthorizationRequestEndpoint == nil {
		return nil, fmt.Errorf("PAR endpoint not found in metadata")
	}

	parEndpoint := *metadata.PushedAuthorizationRequestEndpoint

	// Create DPoP proof for PAR request
	dpopProof, err := dpopKeyPair.CreateDPoPProof("POST", parEndpoint, nil, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create DPoP proof: %w", err)
	}

	// Build PAR request parameters
	data := url.Values{}
	data.Set("response_type", "code")
	data.Set("client_id", c.clientID)
	data.Set("code_challenge", pkce.CodeChallenge)
	data.Set("code_challenge_method", "S256")
	data.Set("scope", "atproto transition:generic")
	data.Set("login_hint", handle)

	req, err := http.NewRequestWithContext(ctx, "POST", parEndpoint, bytes.NewBufferString(data.Encode()))
	if err != nil {
		return nil, fmt.Errorf("failed to create PAR request: %w", err)
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("DPoP", dpopProof)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("PAR request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("PAR failed with status %d: %s", resp.StatusCode, string(body))
	}

	var parResponse PARResponse
	if err := json.NewDecoder(resp.Body).Decode(&parResponse); err != nil {
		return nil, fmt.Errorf("failed to decode PAR response: %w", err)
	}

	return &parResponse, nil
}

// BuildAuthorizationURL builds authorization URL using PAR response
func (c *AtProtoOAuthClient) BuildAuthorizationURL(
	metadata *OAuthServerMetadata,
	parResponse *PARResponse,
) string {
	return fmt.Sprintf(
		"%s?client_id=%s&request_uri=%s",
		metadata.AuthorizationEndpoint,
		url.QueryEscape(c.clientID),
		url.QueryEscape(parResponse.RequestURI),
	)
}

// ExchangeCodeForTokens exchanges authorization code for tokens using DPoP
func (c *AtProtoOAuthClient) ExchangeCodeForTokens(
	ctx context.Context,
	metadata *OAuthServerMetadata,
	code string,
	codeVerifier string,
	dpopKeyPair *DPoPKeyPair,
	redirectURI string,
) (*TokenResponse, error) {
	return c.exchangeCodeWithNonce(ctx, metadata, code, codeVerifier, dpopKeyPair, redirectURI, nil)
}

// exchangeCodeWithNonce exchanges code with optional DPoP nonce
func (c *AtProtoOAuthClient) exchangeCodeWithNonce(
	ctx context.Context,
	metadata *OAuthServerMetadata,
	code string,
	codeVerifier string,
	dpopKeyPair *DPoPKeyPair,
	redirectURI string,
	nonce *string,
) (*TokenResponse, error) {
	// Create DPoP proof for token request
	dpopProof, err := dpopKeyPair.CreateDPoPProof("POST", metadata.TokenEndpoint, nonce, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create DPoP proof: %w", err)
	}

	// Build token request parameters
	data := url.Values{}
	data.Set("grant_type", "authorization_code")
	data.Set("code", code)
	data.Set("redirect_uri", redirectURI)
	data.Set("client_id", c.clientID)
	data.Set("code_verifier", codeVerifier)

	req, err := http.NewRequestWithContext(ctx, "POST", metadata.TokenEndpoint, bytes.NewBufferString(data.Encode()))
	if err != nil {
		return nil, fmt.Errorf("failed to create token request: %w", err)
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("DPoP", dpopProof)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("token request failed: %w", err)
	}
	defer resp.Body.Close()

	// Extract DPoP nonce if present (for retry)
	dpopNonce := resp.Header.Get("dpop-nonce")

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)

		// If we got a nonce and 401, retry with nonce
		if resp.StatusCode == http.StatusUnauthorized && dpopNonce != "" && nonce == nil {
			return c.exchangeCodeWithNonce(ctx, metadata, code, codeVerifier, dpopKeyPair, redirectURI, &dpopNonce)
		}

		return nil, fmt.Errorf("token exchange failed with status %d: %s", resp.StatusCode, string(body))
	}

	var tokenResp TokenResponse
	if err := json.NewDecoder(resp.Body).Decode(&tokenResp); err != nil {
		return nil, fmt.Errorf("failed to decode token response: %w", err)
	}

	// Calculate expiration time
	tokenResp.ExpiresAt = time.Now().Add(time.Duration(tokenResp.ExpiresIn) * time.Second)

	return &tokenResp, nil
}

// RefreshToken refreshes access token using refresh token with DPoP
func (c *AtProtoOAuthClient) RefreshToken(
	ctx context.Context,
	metadata *OAuthServerMetadata,
	refreshToken string,
	dpopKeyPair *DPoPKeyPair,
) (*TokenResponse, error) {
	// Create DPoP proof for token refresh
	dpopProof, err := dpopKeyPair.CreateDPoPProof("POST", metadata.TokenEndpoint, nil, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create DPoP proof: %w", err)
	}

	// Build refresh request parameters
	data := url.Values{}
	data.Set("grant_type", "refresh_token")
	data.Set("refresh_token", refreshToken)
	data.Set("client_id", c.clientID)

	req, err := http.NewRequestWithContext(ctx, "POST", metadata.TokenEndpoint, bytes.NewBufferString(data.Encode()))
	if err != nil {
		return nil, fmt.Errorf("failed to create refresh request: %w", err)
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("DPoP", dpopProof)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("token refresh failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("token refresh failed with status %d: %s", resp.StatusCode, string(body))
	}

	var tokenResp TokenResponse
	if err := json.NewDecoder(resp.Body).Decode(&tokenResp); err != nil {
		return nil, fmt.Errorf("failed to decode refresh response: %w", err)
	}

	// Calculate expiration time
	tokenResp.ExpiresAt = time.Now().Add(time.Duration(tokenResp.ExpiresIn) * time.Second)

	return &tokenResp, nil
}

// MakeAuthenticatedRequest makes an authenticated API request with DPoP
func (c *AtProtoOAuthClient) MakeAuthenticatedRequest(
	ctx context.Context,
	method string,
	apiURL string,
	accessToken string,
	dpopKeyPair *DPoPKeyPair,
	body io.Reader,
) (*http.Response, error) {
	// Calculate access token hash for DPoP
	ath := CalculateAccessTokenHash(accessToken)

	// Create DPoP proof
	dpopProof, err := dpopKeyPair.CreateDPoPProof(method, apiURL, nil, &ath)
	if err != nil {
		return nil, fmt.Errorf("failed to create DPoP proof: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, method, apiURL, body)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Authorization", fmt.Sprintf("DPoP %s", accessToken))
	req.Header.Set("DPoP", dpopProof)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("authenticated request failed: %w", err)
	}

	return resp, nil
}

File: rust-server/src/auth/atproto_oauth.rs
use super::dpop::{calculate_access_token_hash, DPoPKeyPair};
use super::{AuthError, PKCEParams, TokenResponse};
use anyhow::{Context, Result};
use reqwest;
use serde::{Deserialize, Serialize};
use std::time::{Duration, SystemTime};

/// AT Protocol OAuth client with full DPoP support
pub struct AtProtoOAuthClient {
    client_id: String,
    http_client: reqwest::Client,
}

impl AtProtoOAuthClient {
    /// Create a new AT Protocol OAuth client
    pub fn new(client_id: String) -> Self {
        Self {
            client_id,
            http_client: reqwest::Client::builder()
                .timeout(Duration::from_secs(30))
                .build()
                .unwrap(),
        }
    }
    
    /// Discover OAuth server metadata for a PDS following AT Protocol spec
    pub async fn discover_metadata(&self, pds_url: &str) -> Result<OAuthServerMetadata> {
        // Step 1: Get protected resource metadata from PDS to find authorization server
        let protected_resource_url = format!("{}/.well-known/oauth-protected-resource", pds_url);
        
        // Add timeout for protected resource discovery
        let timeout_duration = std::time::Duration::from_secs(10);
        let resp = tokio::time::timeout(
            timeout_duration,
            self.http_client.get(&protected_resource_url).send()
        ).await;
        
        let resp = match resp {
            Ok(Ok(r)) => r,
            Ok(Err(e)) => {
                anyhow::bail!(
                    "Failed to fetch OAuth protected resource metadata from {}: {}. The server may not support AT Protocol OAuth yet. Please try --method password instead.",
                    protected_resource_url, e
                );
            }
            Err(_) => {
                anyhow::bail!(
                    "OAuth protected resource discovery timed out after 10 seconds at {}. The server may not support AT Protocol OAuth yet. Please use --method password instead.",
                    pds_url
                );
            }
        };
        
        if !resp.status().is_success() {
            let status = resp.status();
            let body = resp.text().await.unwrap_or_default();
            anyhow::bail!(
                "OAuth protected resource discovery failed with status {} from {}. Response: {}. The server may not support AT Protocol OAuth yet. Please try --method password instead.",
                status, protected_resource_url, body
            );
        }
        
        #[derive(serde::Deserialize)]
        struct ProtectedResourceMetadata {
            authorization_servers: Vec<String>,
        }
        
        let protected_resource: ProtectedResourceMetadata = resp.json().await
            .context(format!("Failed to decode OAuth protected resource metadata from {}", protected_resource_url))?;
        
        if protected_resource.authorization_servers.is_empty() {
            anyhow::bail!(
                "No authorization servers found in protected resource metadata from {}. The server may not support AT Protocol OAuth yet. Please try --method password instead.",
                protected_resource_url
            );
        }
        
        // Step 2: Get authorization server metadata from the first authorization server
        let auth_server_url = &protected_resource.authorization_servers[0];
        let metadata_url = format!("{}/.well-known/oauth-authorization-server", auth_server_url);
        
        let resp2 = tokio::time::timeout(
            timeout_duration,
            self.http_client.get(&metadata_url).send()
        ).await;
        
        let resp2 = match resp2 {
            Ok(Ok(r)) => r,
            Ok(Err(e)) => {
                anyhow::bail!(
                    "Failed to fetch OAuth authorization server metadata from {}: {}. The server may not support AT Protocol OAuth yet. Please try --method password instead.",
                    metadata_url, e
                );
            }
            Err(_) => {
                anyhow::bail!(
                    "OAuth authorization server metadata discovery timed out after 10 seconds at {}. The server may not support AT Protocol OAuth yet. Please use --method password instead.",
                    auth_server_url
                );
            }
        };
        
        if !resp2.status().is_success() {
            let status = resp2.status();
            let body = resp2.text().await.unwrap_or_default();
            anyhow::bail!(
                "OAuth authorization server metadata discovery failed with status {} from {}. Response: {}. The server may not support AT Protocol OAuth yet. Please try --method password instead.",
                status, metadata_url, body
            );
        }
        
        let metadata: OAuthServerMetadata = resp2.json().await
            .context(format!("Failed to decode OAuth authorization server metadata from {}", metadata_url))?;
        Ok(metadata)
    }
    
    /// Send a Pushed Authorization Request (PAR)
    pub async fn send_par(
        &self,
        metadata: &OAuthServerMetadata,
        pkce: &PKCEParams,
        dpop_keypair: &DPoPKeyPair,
        handle: &str,
    ) -> Result<PARResponse, AuthError> {
        let par_endpoint = metadata.pushed_authorization_request_endpoint.as_ref()
            .ok_or_else(|| AuthError::OAuthError("PAR endpoint not found".to_string()))?;
        
        // Create DPoP proof for PAR request
        let dpop_proof = dpop_keypair.create_dpop_proof(
            "POST",
            par_endpoint,
            None,
            None,
        ).map_err(|e| AuthError::OAuthError(format!("Failed to create DPoP proof: {}", e)))?;
        
        // Build PAR request
        let params = [
            ("response_type", "code"),
            ("client_id", &self.client_id),
            ("code_challenge", &pkce.code_challenge),
            ("code_challenge_method", "S256"),
            ("scope", "atproto transition:generic"),
            ("login_hint", handle),
        ];
        
        let resp = self.http_client
            .post(par_endpoint)
            .header("DPoP", dpop_proof)
            .header("Content-Type", "application/x-www-form-urlencoded")
            .form(&params)
            .send()
            .await
            .map_err(|e| AuthError::NetworkError(format!("PAR request failed: {}", e)))?;
        
        let status = resp.status();
        if !status.is_success() {
            let body = resp.text().await.unwrap_or_default();
            return Err(AuthError::OAuthError(format!(
                "PAR failed with status {}: {}",
                status,
                body
            )));
        }
        
        let par_response: PARResponse = resp
            .json()
            .await
            .map_err(|e| AuthError::ParseError(format!("Failed to decode PAR response: {}", e)))?;
        
        Ok(par_response)
    }
    
    /// Build authorization URL using PAR response
    pub fn build_authorization_url(
        &self,
        metadata: &OAuthServerMetadata,
        par_response: &PARResponse,
    ) -> String {
        format!(
            "{}?client_id={}&request_uri={}",
            metadata.authorization_endpoint,
            urlencoding::encode(&self.client_id),
            urlencoding::encode(&par_response.request_uri)
        )
    }
    
    /// Exchange authorization code for tokens using DPoP
    pub async fn exchange_code_for_tokens(
        &self,
        metadata: &OAuthServerMetadata,
        code: &str,
        code_verifier: &str,
        dpop_keypair: &DPoPKeyPair,
        redirect_uri: &str,
    ) -> Result<TokenResponse, AuthError> {
        // Create DPoP proof for token request
        let dpop_proof = dpop_keypair.create_dpop_proof(
            "POST",
            &metadata.token_endpoint,
            None,
            None,
        ).map_err(|e| AuthError::OAuthError(format!("Failed to create DPoP proof: {}", e)))?;
        
        let params = [
            ("grant_type", "authorization_code"),
            ("code", code),
            ("redirect_uri", redirect_uri),
            ("client_id", &self.client_id),
            ("code_verifier", code_verifier),
        ];
        
        let resp = self.http_client
            .post(&metadata.token_endpoint)
            .header("DPoP", dpop_proof)
            .header("Content-Type", "application/x-www-form-urlencoded")
            .form(&params)
            .send()
            .await
            .map_err(|e| AuthError::NetworkError(format!("Token request failed: {}", e)))?;
        
        let status = resp.status();
        
        // Extract DPoP nonce if present (for retry)
        let dpop_nonce = resp.headers()
            .get("dpop-nonce")
            .and_then(|v| v.to_str().ok())
            .map(|s| s.to_string());
        
        if !status.is_success() {
            let body = resp.text().await.unwrap_or_default();
            
            // If we got a nonce and 401, retry with nonce
            if status.as_u16() == 401 && dpop_nonce.is_some() {
                return self.exchange_code_for_tokens_with_nonce(
                    metadata,
                    code,
                    code_verifier,
                    dpop_keypair,
                    redirect_uri,
                    dpop_nonce.as_deref(),
                ).await;
            }
            
            return Err(AuthError::OAuthError(format!(
                "Token exchange failed with status {}: {}",
                status,
                body
            )));
        }
        
        let mut token_resp: TokenResponse = resp
            .json()
            .await
            .map_err(|e| AuthError::ParseError(format!("Failed to decode token response: {}", e)))?;
        
        // Calculate expiration time
        token_resp.expires_at = SystemTime::now() + Duration::from_secs(token_resp.expires_in as u64);
        
        Ok(token_resp)
    }
    
    /// Exchange code with DPoP nonce (retry)
    async fn exchange_code_for_tokens_with_nonce(
        &self,
        metadata: &OAuthServerMetadata,
        code: &str,
        code_verifier: &str,
        dpop_keypair: &DPoPKeyPair,
        redirect_uri: &str,
        nonce: Option<&str>,
    ) -> Result<TokenResponse, AuthError> {
        // Create DPoP proof with nonce
        let dpop_proof = dpop_keypair.create_dpop_proof(
            "POST",
            &metadata.token_endpoint,
            nonce,
            None,
        ).map_err(|e| AuthError::OAuthError(format!("Failed to create DPoP proof: {}", e)))?;
        
        let params = [
            ("grant_type", "authorization_code"),
            ("code", code),
            ("redirect_uri", redirect_uri),
            ("client_id", &self.client_id),
            ("code_verifier", code_verifier),
        ];
        
        let resp = self.http_client
            .post(&metadata.token_endpoint)
            .header("DPoP", dpop_proof)
            .header("Content-Type", "application/x-www-form-urlencoded")
            .form(&params)
            .send()
            .await
            .map_err(|e| AuthError::NetworkError(format!("Token request failed: {}", e)))?;
        
        let status = resp.status();
        if !status.is_success() {
            let body = resp.text().await.unwrap_or_default();
            return Err(AuthError::OAuthError(format!(
                "Token exchange failed with status {}: {}",
                status,
                body
            )));
        }
        
        let mut token_resp: TokenResponse = resp
            .json()
            .await
            .map_err(|e| AuthError::ParseError(format!("Failed to decode token response: {}", e)))?;
        
        // Calculate expiration time
        token_resp.expires_at = SystemTime::now() + Duration::from_secs(token_resp.expires_in as u64);
        
        Ok(token_resp)
    }
    
    /// Refresh access token using refresh token with DPoP
    pub async fn refresh_token(
        &self,
        metadata: &OAuthServerMetadata,
        refresh_token: &str,
        dpop_keypair: &DPoPKeyPair,
    ) -> Result<TokenResponse, AuthError> {
        // Create DPoP proof for token refresh
        let dpop_proof = dpop_keypair.create_dpop_proof(
            "POST",
            &metadata.token_endpoint,
            None,
            None,
        ).map_err(|e| AuthError::OAuthError(format!("Failed to create DPoP proof: {}", e)))?;
        
        let params = [
            ("grant_type", "refresh_token"),
            ("refresh_token", refresh_token),
            ("client_id", &self.client_id),
        ];
        
        let resp = self.http_client
            .post(&metadata.token_endpoint)
            .header("DPoP", dpop_proof)
            .header("Content-Type", "application/x-www-form-urlencoded")
            .form(&params)
            .send()
            .await
            .map_err(|e| AuthError::NetworkError(format!("Token refresh failed: {}", e)))?;
        
        let status = resp.status();
        if !status.is_success() {
            let body = resp.text().await.unwrap_or_default();
            return Err(AuthError::OAuthError(format!(
                "Token refresh failed with status {}: {}",
                status,
                body
            )));
        }
        
        let mut token_resp: TokenResponse = resp
            .json()
            .await
            .map_err(|e| AuthError::ParseError(format!("Failed to decode refresh response: {}", e)))?;
        
        // Calculate expiration time
        token_resp.expires_at = SystemTime::now() + Duration::from_secs(token_resp.expires_in as u64);
        
        Ok(token_resp)
    }
    
    /// Make an authenticated API request with DPoP
    pub async fn make_authenticated_request(
        &self,
        method: &str,
        url: &str,
        access_token: &str,
        dpop_keypair: &DPoPKeyPair,
    ) -> Result<reqwest::Response> {
        // Calculate access token hash for DPoP
        let ath = calculate_access_token_hash(access_token);
        
        // Create DPoP proof
        let dpop_proof = dpop_keypair.create_dpop_proof(
            method,
            url,
            None,
            Some(&ath),
        )?;
        
        // Make request
        let req = match method {
            "GET" => self.http_client.get(url),
            "POST" => self.http_client.post(url),
            "PUT" => self.http_client.put(url),
            "DELETE" => self.http_client.delete(url),
            _ => anyhow::bail!("Unsupported HTTP method: {}", method),
        };
        
        let resp = req
            .header("Authorization", format!("DPoP {}", access_token))
            .header("DPoP", dpop_proof)
            .send()
            .await?;
        
        Ok(resp)
    }
}

/// OAuth server metadata from .well-known discovery
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OAuthServerMetadata {
    pub issuer: String,
    pub authorization_endpoint: String,
    pub token_endpoint: String,
    pub pushed_authorization_request_endpoint: Option<String>,
    pub dpop_signing_alg_values_supported: Option<Vec<String>>,
    #[serde(default)]
    pub response_types_supported: Vec<String>,
    #[serde(default)]
    pub grant_types_supported: Vec<String>,
    #[serde(default)]
    pub code_challenge_methods_supported: Vec<String>,
}

/// PAR (Pushed Authorization Request) response
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PARResponse {
    pub request_uri: String,
    pub expires_in: i64,
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_create_atproto_client() {
        let client = AtProtoOAuthClient::new("test-client".to_string());
        assert_eq!(client.client_id, "test-client");
    }
    
    #[test]
    fn test_build_authorization_url() {
        let client = AtProtoOAuthClient::new("test-client".to_string());
        let metadata = OAuthServerMetadata {
            issuer: "https://example.com".to_string(),
            authorization_endpoint: "https://example.com/authorize".to_string(),
            token_endpoint: "https://example.com/token".to_string(),
            pushed_authorization_request_endpoint: Some("https://example.com/par".to_string()),
            dpop_signing_alg_values_supported: Some(vec!["ES256".to_string()]),
            response_types_supported: vec!["code".to_string()],
            grant_types_supported: vec!["authorization_code".to_string()],
            code_challenge_methods_supported: vec!["S256".to_string()],
        };
        
        let par_response = PARResponse {
            request_uri: "urn:ietf:params:oauth:request_uri:test123".to_string(),
            expires_in: 90,
        };
        
        let url = client.build_authorization_url(&metadata, &par_response);
        assert!(url.contains("request_uri="));
        assert!(url.contains("client_id="));
    }
}
